### ISSUE TITLE
Where is the !command handler located

### ISSUE THREAD

**Issue by jules-datascientest** [ipython/ipython#13308]
Hi, I've missasked my question yesterday on ipykernel which is why I'm asking again here (https://github.com/ipython/ipykernel/issues/802)

Where is the handling for `!xxx` commands (like pip) are handled ? I'd need to modify it to block pip installs on jupyter notebooks.

Thanks a lot

**Comment by minrk** [comment]
Modifying `!` to block pip isn't going to do much to prevent pip installs because there are _many_ other ways to do it (`os.system`, `subprocess.run`, `import pip`, etc.). So it depends a bit on what your goal is. But it's a shortcut to `get_ipython().system(...)` which you can override with your own implementation (anyone running a notebook can also override it to anything else they want, so this wouldn't be a security measure. It might let you provide a useful message in the simple case, though, if that's what you are after).

**Comment by jules-datascientest** [comment]
Sadly not it isn't what I'm after but thanks for your reply !

Our objective is to completely block pip installs from notebooks, but I think our best shot would be to block it on the user level directly, rather than on jupyter.

**Comment by bollwyvl** [comment]
Yeah, hacking in the kernel isn't going to do the thing in any meaningful way. Some options would be:
- forcibly remove pip binaries
  - but it's not doing _that_ much, and it's easy to work around
- make the environment read-only on disk
  - but there are ways to add additional things to the path
- block PyPI
  - helps, but there are so many other ways to get more software

**Comment by Carreau** [comment]
If you want to block pip you are going to need to block Python, as you can always to `python -m pip`. 

As for your exact questions, `!` is in the end handled by system_piped and system_raw. 

https://github.com/ipython/ipython/blob/c13d3165790fa6af33efe89fb08b6c4ae0d5bd13/IPython/core/interactiveshell.py#L2390-L2480

You also want to look at sx/sc

https://github.com/ipython/ipython/blob/c13d3165790fa6af33efe89fb08b6c4ae0d5bd13/IPython/core/magics/osm.py#L681-L734

https://github.com/ipython/ipython/blob/c13d3165790fa6af33efe89fb08b6c4ae0d5bd13/IPython/core/magics/osm.py#L735-L736
